### PR TITLE
Feature maintenance page

### PR DESCRIPTION
**Pull request by michalpiatek**

### CODE REVIEW

File: src/test/java/uk.gov.cshr/controller/MaintenanceControllerTest.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+    @Test
+    public void shouldRedirectToMaintenance() throws Exception {
+        mockMvc
+            .perform(get("/login"))
+            .andExpect(redirectedUrl("/maintenance"));
+    }
+
+    @Test
+    public void assetsShouldRedirectToMaintenance() throws Exception {
```
Comment (charlesluokb) [inline src/test/java/uk.gov.cshr/controller/MaintenanceControllerTest.java]: can we add another scenario `oauthShouldRedirectToMaintenance()`?

File: src/test/java/uk/gov/cshr/controller/MaintenanceControllerDisabledTest.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    @Autowired
+    private Helper helper;
+
+    @Before
+    public void overrideFilters() throws IllegalAccessException {
+        helper.initAppInsighsFilter(mockMvc);
+    }
+
+    @Test
+    public void shouldRedirectToMaintenance() throws Exception {
```
Comment (charlesluokb) [inline src/test/java/uk/gov/cshr/controller/MaintenanceControllerDisabledTest.java]: `should` -> `shouldNot`?

File: src/test/java/uk/gov/cshr/controller/MaintenanceControllerDisabledTest.java [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.redirectedUrl;
+
+@RunWith(SpringRunner.class)
+@SpringBootTest
+@AutoConfigureMockMvc
+@Transactional
+@TestPropertySource(properties = {"MAINTENANCE_ENABLED = false"})
+public class MaintenanceControllerDisabledTest {
```
Comment (charlesluokb) [inline src/test/java/uk/gov/cshr/controller/MaintenanceControllerDisabledTest.java]: not sure if adding another test scenario `shouldNotRedirectToMaintenanceWhenCallingOauth` is an overkill?

File: src/test/java/uk/gov/cshr/AppInsightsHelper/Helper.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+import org.springframework.stereotype.Component;
+import org.springframework.test.web.servlet.MockMvc;
+
+import javax.servlet.Filter;
+import javax.servlet.FilterConfig;
+import javax.servlet.ServletContext;
+import java.util.Enumeration;
+
+@Component
+public class Helper {
```
Comment (charlesluokb) [inline src/test/java/uk/gov/cshr/AppInsightsHelper/Helper.java]: just a thought. `FilterHelper` can have better meaning than just `Helper`. Nice to see it is annotated as a spring bean though.

File: src/main/resources/templates/maintenance.html [hunk 5]
```diff
@@ -0,0 +1,26 @@
+<page th:replace="fragments/maintenance-layout :: page(~{::title},~{::content})">
```
Comment (charlesluokb) [inline src/main/resources/templates/maintenance.html]: not sure if the contents of maintenance page have been approved by the client. Also if we can also provide info about when the site will be available?

File: src/main/java/uk/gov/cshr/filter/MaintenanceFilter.java [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+        return !maintenanceEnabled &&
+                isMaintenaceRequest(httpRequest);
+    }
+
+    private boolean isMaintenaceRequest(HttpServletRequest httpRequest) {
+        return httpRequest.getRequestURI().equalsIgnoreCase(MAINTENANCE_URI);
+    }
+
+    private boolean isWhiteListedRequest(HttpServletRequest httpRequest) {
+        return httpRequest.getRequestURI().startsWith(ASSETS_PATH) ||
```
Comment (charlesluokb) [inline src/main/java/uk/gov/cshr/filter/MaintenanceFilter.java]: why we whitelist `/assets` path?

File: src/test/java/uk.gov.cshr/controller/MaintenanceControllerTest.java (hunk 1 above)
Comment (michalpiatek) [reply src/test/java/uk.gov.cshr/controller/MaintenanceControllerTest.java]: sure

File: src/test/java/uk/gov/cshr/controller/MaintenanceControllerDisabledTest.java (hunk 3 above)
Comment (michalpiatek) [reply src/test/java/uk/gov/cshr/controller/MaintenanceControllerDisabledTest.java]: can remove

File: src/test/java/uk/gov/cshr/controller/MaintenanceControllerDisabledTest.java (hunk 2 above)
Comment (michalpiatek) [reply src/test/java/uk/gov/cshr/controller/MaintenanceControllerDisabledTest.java]: will change

File: src/test/java/uk/gov/cshr/AppInsightsHelper/Helper.java (hunk 4 above)
Comment (michalpiatek) [reply src/test/java/uk/gov/cshr/AppInsightsHelper/Helper.java]: yip

File: src/main/java/uk/gov/cshr/filter/MaintenanceFilter.java (hunk 6 above)
Comment (michalpiatek) [reply src/main/java/uk/gov/cshr/filter/MaintenanceFilter.java]: as the filter handles all requests - hope that explains it?

File: src/main/resources/templates/maintenance.html (hunk 5 above)
Comment (michalpiatek) [reply src/main/resources/templates/maintenance.html]: good question - I believe I took the content from what was already used as maintenance page - can be changed if needs be of course

File: src/main/java/uk/gov/cshr/filter/MaintenanceFilter.java (hunk 6 above)
Comment (charlesluokb) [reply src/main/java/uk/gov/cshr/filter/MaintenanceFilter.java]: as per the settings in `UserSecurityConfig` 
`.antMatchers("/login", "/oauth/logout", "/webjars/**", "/assets/**", "/signup/**", "/reset/**",  "/account/passwordUpdated").permitAll()`
why paths like /reset or /webjars are not whitelisted?

Comment (michalpiatek) [reply src/main/java/uk/gov/cshr/filter/MaintenanceFilter.java]: cause we only need /assets/* and /oauth/* for the override to work. It is a whitelist for the maintenance mode override not whitelist authentication/authorisation wise :)

**Comment by m24murray** [comment]
Closing - has been replaced by https://github.com/Civil-Service-Human-Resources/identity-service/pull/113

**Pull request closed without merging**